### DISCUSSION TITLE
mem rise up slowly

### DISCUSSION THREAD

**Discussion by zcgecho** [emqx/emqx#4609] [Q&A]
i test emqx on centos7 for v4.2 in standalone mode recently, and i find that its mem rise up  slowly.
test conditon:
300 messages published per second and about 1000 clients(pub client and sub client) connected ,the mem takes 250M to 400M in 24hours later. and i still plan to test on an on to see  how much mem it takes to the end.
i know emqx build on erlang and GC is atomatically occured.

**Comment by tigercl** [comment]
@zcgecho It may be that the consumer is too late to spend, causing the messages to slowly accumulate in EMQ X.

**Comment by qzhuyan** [comment]
@zcgecho 
It could be some erlang vm's cache wasn't returned to the OS for handling spike load.

to verify that, could you attach the emqx's remote console and send us the output of the following command?

```
# in linux shell
emqx remote_console
```
then in the console:

```
# in emqx console
 erlang:memory().
 rp(recon_alloc:allocators()).
```

**Comment by zcgecho** [comment]
thanks very much  ,i will check it .

**Comment by zcgecho** [comment]
> @zcgecho
> It could be some erlang vm's cache wasn't returned to the OS for handling spike load.
> 
> to verify that, could you attach the emqx's remote console and send us the output of the following command?
> 
> ```
> # in linux shell
> emqx remote_console
> ```
> 
> then in the console:
> 
> ```
> # in emqx console
>  erlang:memory().
>  rp(recon_alloc:allocators()).
> ```



i tese just in standalone ,so not use akka. it seems in standalone mod the cmd "emqx remote_console " does not works.

./emqx remote_console
Protocol 'ekka': register/listen error: eaddrinuse

**Comment by zcgecho** [comment]
> I

thanks for your reply.

i boot sub program firstly and then push msg to a certain topic  which can be consumed by certain subscribers definetely.
so i don't think msg will be blocked， and also that msg is not retained in broker(retain flag false).

by the way ,i found a strange thing that the mem seems to be stable and normal if msg is a small size(json about 100 byte)，but  mem rise up slowly if msg size is much bigger(json about 10k).

**Comment by zmstone** [comment]
> > @zcgecho
> > It could be some erlang vm's cache wasn't returned to the OS for handling spike load.
> > to verify that, could you attach the emqx's remote console and send us the output of the following command?
> > ```
> > # in linux shell
> > emqx remote_console
> > ```
> > 
> > 
> > then in the console:
> > ```
> > # in emqx console
> >  erlang:memory().
> >  rp(recon_alloc:allocators()).
> > ```
> 
> i tese just in standalone ,so not use akka. it seems in standalone mod the cmd "emqx remote_console " does not works.
> 
> ./emqx remote_console
> Protocol 'ekka': register/listen error: eaddrinuse

please try  the attach command instead of remote_console

**Comment by zcgecho** [comment]
> > > @zcgecho
> > > It could be some erlang vm's cache wasn't returned to the OS for handling spike load.
> > > to verify that, could you attach the emqx's remote console and send us the output of the following command?
> > > ```
> > > # in linux shell
> > > emqx remote_console
> > > ```
> > > 
> > > 
> > > then in the console:
> > > ```
> > > # in emqx console
> > >  erlang:memory().
> > >  rp(recon_alloc:allocators()).
> > > ```
> > 
> > 
> > i tese just in standalone ,so not use akka. it seems in standalone mod the cmd "emqx remote_console " does not works.
> > ./emqx remote_console
> > Protocol 'ekka': register/listen error: eaddrinuse
> 
> please try the attach command instead of remote_console

i reproduce this issue when msg published to broker is very bigger(json about 17k) with high frequncy. then i use ./emqx attach cmd error as follow shows .
--------------------
properties => #{},proto_ver => 4,protocol => mqtt,username => <<"admin">>})
2021-04-13 08:18:11.852 [warning] <<"Sub_Clientid_10002">>@10.130.212.71:64225 [Session] Dropped msg due to mqueue is full: Message(Id=^@^E¿Ï 
ÛôBZ^@^B^X^@^A, QoS=2, Topic=/topic/10339, From=<<"pub_Clientid_10183">>, Flags=[], Headers=#{peerhost => {10,130,212,71},
--------------------------------
still in the shell later,warnig shows as follow, it seems queue is full
-----------------
properties => #{},proto_ver => 4,protocol => mqtt,username => <<"admin">>})
2021-04-13 08:18:11.456 [warning] <<"Sub_Clientid_10019">>@10.130.212.71:64283 [Session] Dropped msg due to mqueue is full: Message(Id=^@^E¿Ï^EìKôBZ^@^B^C^@^A, QoS=2, Topic=/topic/10333, From=<<"pub_Clientid_10177">>, Flags=[], Headers=#{peerhost => {10,130,212,71},
------

and last  emqx process crashed and find no helpful info in crash.dump file.

**Comment by zmstone** [comment]
what you are seeing is the logs being sent to the attached pipe (console),
the logs should also be written to log files (if configured).

yes, this log implies message are being dropped due to the internal message buffer (queue) is full.
which is typically caused either due to publishers being aggressive or subscribers being slow.

I would not worry about the slow increase of RAM usage
unless hitting the limit and getting OOM killed by os or other system tools.
Erlang garbage collection is rather not aggressive, causing memory to be fragmented.

Nonetheless, the commands suggested by @qzhuyan should give us more insights
if you can execute the commands when attached to the console.
```
 erlang:memory().
 rp(recon_alloc:allocators()).
```

You can configure `log.to` in `emqx.conf` to `file` (not `both`) to avoid the Erlang console getting flooded by logs.

**Comment by qzhuyan** [comment]
if @zcgecho sees the "Dropped msg " warnings after attaching to the console, doesn't it mean it is just happening?

Could that be caused by some slow subscribers or even offline subscribers?

**Comment by zmstone** [comment]
> if @zcgecho sees the "Dropped msg " warnings after attaching to the console, doesn't it mean it is just happening?
> 
> Could that be caused by some slow subscribers or even offline subscribers?

Yes it’s ongoing event which is not surprising given the high message rate.
Offline clients with persistent session could be another possibility yes
